File: src/main/java/pl/training/workshop/m7/s03_breakresponsibilities/step3/TicketPricer.java
package pl.training.workshop.m7.s03_breakresponsibilities.step3;

import java.math.BigDecimal;

import pl.training.workshop.m7.s03_breakresponsibilities.BookingRequest;

/** Krok 2: Extract Class - cennik ma jednego właściciela i zwraca wynik zamiast dwóch zmiennych. */
final class TicketPricer {
    Pricing price(BookingRequest request) {
        BigDecimal base = switch (request.format()) {
            case "IMAX" -> new BigDecimal("40.00");
            case "3D" -> new BigDecimal("32.00");
            default -> new BigDecimal("25.00");
        };
        BigDecimal total = BigDecimal.ZERO;
        int vipSeats = 0;
        for (String seat : request.seats()) {
            total = total.add(base);
            if (Integer.parseInt(seat.substring(1)) >= 10) {
                total = total.add(new BigDecimal("10.00"));
                vipSeats++;
            }
        }
        return new Pricing(total, vipSeats);
    }
}

File: typescript/src/workshop/m7/s03_breakresponsibilities/step2/TicketPricer.ts
import { Decimal } from 'decimal.js';

import type { BookingRequest } from '../BookingRequest.js';
import { Pricing } from './Pricing.js';

/** Krok 2: Extract Class - cennik ma jednego właściciela i zwraca wynik zamiast dwóch zmiennych. */
export class TicketPricer {
  price(request: BookingRequest): Pricing {
    let base: Decimal;
    switch (request.format) {
      case 'IMAX': base = new Decimal('40.00'); break;
      case '3D': base = new Decimal('32.00'); break;
      default: base = new Decimal('25.00');
    }
    let total = new Decimal(0);
    let vipSeats = 0;
    for (const seat of request.seats) {
      total = total.plus(base);
      if (Number.parseInt(seat.substring(1), 10) >= 10) {
        total = total.plus(new Decimal('10.00'));
        vipSeats++;
      }
    }
    return new Pricing(total, vipSeats);
  }
}
